### DISCUSSION TITLE
NativeActivity is destroyed when a USB device is attached or detached.

### DISCUSSION THREAD

**Discussion by IllusionMan1212** [android/ndk#2055] [Q&A] [answered]
As the title suggests, is there any solution or workaround for when this happens?

**Comment by DanAlbert** [comment]
Maybe another user listening here knows an answer, but if you want a fix you probably need to file an OS bug. `NativeActivity` is part of the OS.

One thing you might try would be switching to `GameActivity`: https://developer.android.com/games/agdk/game-activity. If the OS is sending the app destroy event when it shouldn't be though, there's nothing that can be done about this from the app side.

**Reply by IllusionMan1212** [reply]
I was hesitant to file a bug because I wasn't sure if this is intended behavior or not. When the USB device is attached/detached the application receives PAUSE, TERM_WINDOW, STOP, and INPUT_CHANGED commands which is normal from my understanding because a USB (de)attachment is a configuration change, but what's abnormal in my opinion is that `destroyRequested` is set to 1 and no RESUME, START, or INIT_WINDOW commands are received. 

I think `GameActivity` will suffer from the same problem since it extends `NativeActivity`.

Can you point me to the correct place where I should file the OS bug?

**Reply by IllusionMan1212** [reply, marked as answer]
So I decided to try the ndk samples to see if they were affected too, turns out they weren't so it was clear I was doing something wrong. It turns out I was confused on what `destroyRequested` is supposed to signal and thought the `android_main` function will only be called once by the `NativeActivity`, so I never let the app thread exit when it should have. In hindsight it makes total sense that `android_main` will be called every time the `NativeActivity` is destroyed and remade.

Thank you for taking the time to reply.

**Reply by DanAlbert** [reply]
Ah, yeah, that'll do it.

> I think GameActivity will suffer from the same problem since it extends NativeActivity.

Actually it doesn't, it extends `AppCompatActivity`: https://android.googlesource.com/platform/frameworks/opt/gamesdk/+/refs/heads/main/game-activity/src/main/java/com/google/androidgamesdk/GameActivity.java. Even though it's not relevant to this problem, that can help in a lot of other ways.

> Can you point me to the correct place where I should file the OS bug?

In case you need to know in the future, http://b.android.com, and then one of the components in the section titled "platform" (https://source.android.com/docs/setup/contribute/report-bugs#platform). When in doubt, Framework.

**Comment by enh-google** [comment]
do you have a suggested improvement to the `destroyRequested` doc comment?

(i'm assuming that the fix you'd suggest to the `android_main` doc comments is to make it clear that it's not a one-off on _initial_ creation, it's _every_ time.)

**Reply by IllusionMan1212** [reply]
Nothing concrete but maybe append `The android_main function must return to its caller if this is non-zero.` to the existing comment.

I think because I was only writing native code I wasn't thinking about the usual activity lifecycle. I expected STOP/PAUSE and START/RESUME to be at play but it completely slipped my mind that the main function will have to be called more than once when the OS recreates the activity, so maybe making it clear somewhere that the main function is subject to being called multiple times is good too.

**Reply by enh-google** [reply]
does https://android-review.googlesource.com/c/platform/ndk/+/3218918 look better?

**Reply by IllusionMan1212** [reply]
Looks great. Thank you
